runtime/gc_modron_startup/mminit.cpp

```diff
@@ -3017,7 +3017,7 @@ gcInitializeDefaults(J9JavaVM* vm)
 
 		/* We are going to try again -- free any buffer we already have from j9gc_initialize_heap */
 		if  ((loadInfo->loadFlags & FREE_ERROR_STRING) && (NULL != loadInfo->fatalErrorStr)) {
-			j9mem_free_memory(loadInfo->fatalErrorStr);
+			j9mem_free_memory((char *)loadInfo->fatalErrorStr);
```

Comment:
This cast can be avoided as well if `buffer` (in the suggestion above) has a larger scope.

Reply:
I'm a little confused by this suggestion, since this call to `j9mem_free_memory` appears in a different function than the above calls to `j9mem_allocate_memory` that I would be using `buffer` for. Should I define `buffer` in `mminit.h` and initialize it to `NULL`?

Reply:
You're right, I missed the fact that this is a different function, but the idea was to have a different local (of type `char *`) to remember an allocated block that would be freed here if necessary.

As I looked more closely at this function, this is either just dead code: `FREE_ERROR_STRING` is not set earlier in the function, but if that flag was set before the function was called, there are several places that update `fatalErrorStr` without checking if `FREE_ERROR_STRING` is set first. I suggest we want a helper function that handles updating `fatalErrorStr` and that flag properly.

Reply:
Should we make a function that specifically handles freeing `fatalErrorStr`, or one that is more versatile and can handle assignments like the `j9mem_allocate_memory` cases? Would it be a method in [`J9VMDllLoadInfo`](https://github.com/eclipse-openj9/openj9/blob/b03d898a204183ae8594cc07741cc47e44c4d65c/runtime/oti/j9nonbuilder.h#L871) where `fatalErrorStr` lives? I'm an intern on the common optimizer team in Toronto and this would be my first time adding some code (relatively) from scratch, so I want to make sure I understand exactly what should be done.

Reply:
I think we should declare something like this in `vm_api.h` (where related `J9VMDllLoadInfo` functions are declared):
```c
/**
 * Update the fatalErrorStr field of the supplied J9VMDllLoadInfo object,
 * freeing as necessary any, allocated string owned by that object.
 *
 * @param portLib the port library
 * @param info the J9VMDllLoadInfo object
 * @param error the new string (may be NULL)
 * @param errorIsAllocated indicates whether error (if not NULL) must eventually be freed
 */
void setErrorJ9dll(J9PortLibrary *portLib, J9VMDllLoadInfo *info const char *error, BOOLEAN errorIsAllocated);
```
With that we would replace existing assignments
```c
    info->fatalErrorStr = error;
```
with
```c
    setErrorJ9dll(PORTLIB, info, error, TRUE);
```
or
```c
    setErrorJ9dll(PORTLIB, info, error, FALSE);
```
depending on whether `error` must later be freed (or not).

Reply:
Ok thank you for the detailed explanation Keith, I will get on that!